target/arc/fpu-helper.c

```diff
@@ -0,0 +1,938 @@
+
+static uint8_t fpr_width;
+uint8_t vfp_width;
+/* TODO: Maybe is not needed */
+static uint8_t fpr_per_vector_operand;
+static uint8_t vfp_max_length;
+
+void
+init_fpu(CPUARCState *env, bool fp_dp, bool fp_wide, bool fp_hp)
+{
+  uint8_t index1 = ((fp_dp != 0) * 2) + (fp_wide != 0);
+  const uint8_t fpr_widths [] = { 32, 32, 64, 64 };
+  const uint8_t vfp_widths [] = { 32, 64, 64, 128 };
+
+  fpr_width = fpr_widths[index1];
+  vfp_width = vfp_widths[index1];
+
+  fpr_per_vector_operand = (vfp_width / fpr_width);
+  vfp_max_length = fpr_per_vector_operand * (vfp_width / 32) * (fp_hp != 0 ? 2 : 1);
+
+  uint32_t tmp = 0;
+  set_default_nan_mode(0, &env->fp_status);
+  set_float_rounding_mode(float_round_nearest_even, &env->fp_status);
+  set_float_exception_flags(tmp, &env->fp_status);
+
+  env->fp_status_persistent = arc_pack_fpu_status(env);
+}
+
+static inline void check_fpu_raise_exception(CPUARCState *env)
+{
+    env->fp_status_persistent |= arc_pack_fpu_status(env);
+    qemu_log_mask(LOG_UNIMP, "FLAG = 0x%08x\n", (uint32_t) env->fp_status_persistent);
+
+    if((BIT_READ(env->fp_status_persistent, IV)
+        && env->enable_invop_excp != 0)
+       || (BIT_READ(env->fp_status_persistent, DZ)
+	   && env->enable_divzero_excp != 0)) {
+	arc_raise_exception(env, GETPC(), EXCP_EXTENSION);
+    }
+}
+
+#define FLOAT_INST3_HELPERS(NAME, HELPER, SIZE) \
+uint64_t helper_##NAME(CPUARCState *env, uint64_t frs1, uint64_t frs2) \
+{ \
+    uint64_t ret = float##SIZE##_##HELPER(frs1, frs2, &env->fp_status); \
+    check_fpu_raise_exception(env); \
+    return ret; \
+} \
+
+FLOAT_INST3_HELPERS(fdadd, add, 64)
+FLOAT_INST3_HELPERS(fdsub, sub, 64)
+FLOAT_INST3_HELPERS(fdmul, mul, 64)
+FLOAT_INST3_HELPERS(fddiv, div, 64)
+FLOAT_INST3_HELPERS(fdmin, minnum, 64)
+FLOAT_INST3_HELPERS(fdmax, maxnum, 64)
+
+FLOAT_INST3_HELPERS(fsadd, add, 32)
+FLOAT_INST3_HELPERS(fssub, sub, 32)
+FLOAT_INST3_HELPERS(fsmul, mul, 32)
+FLOAT_INST3_HELPERS(fsdiv, div, 32)
+FLOAT_INST3_HELPERS(fsmin, minnum, 32)
+FLOAT_INST3_HELPERS(fsmax, maxnum, 32)
+
+FLOAT_INST3_HELPERS(fhadd, add, 16)
+FLOAT_INST3_HELPERS(fhsub, sub, 16)
+FLOAT_INST3_HELPERS(fhmul, mul, 16)
+FLOAT_INST3_HELPERS(fhdiv, div, 16)
+FLOAT_INST3_HELPERS(fhmin, minnum, 16)
+FLOAT_INST3_HELPERS(fhmax, maxnum, 16)
+
+    /*
+#define FLOAT_SGN_HELPERS(NAME, OPERATION, SIZE) \
+uint64_t helper_##NAME(CPUARCState *env, uint64_t s1, uint64_t s2) \
+{ \
+    return float##SIZE##_set_sign(s1, OPERATION); \
+} \
+
+FLOAT_SGN_HELPERS(fdsgnj,  ( (s2 >> 63) & 1), 64)
+FLOAT_SGN_HELPERS(fdsgnjn, (((s2 >> 63) & 1) == 0), 64)
+FLOAT_SGN_HELPERS(fdsgnjx, ( (s2 >> 63) ^ (s1 >> 63)), 64)
+
+FLOAT_SGN_HELPERS(fssgnj,  ( (s2 >> 31) & 1), 32)
+FLOAT_SGN_HELPERS(fssgnjn, (((s2 >> 31) & 1) == 0), 32)
+FLOAT_SGN_HELPERS(fssgnjx, ( (s2 >> 31) ^ (s1 >> 31)), 32)
+
+FLOAT_SGN_HELPERS(fhsgnj,  ( (s2 >> 15) & 1), 16)
+FLOAT_SGN_HELPERS(fhsgnjn, (((s2 >> 15) & 1) == 0), 16)
+FLOAT_SGN_HELPERS(fhsgnjx, ( (s2 >> 15) ^ (s1 >> 15)), 16)
+*/
```

Comment:
Are u planning to enable these helpers? Please add a either a TODO explaining why are still commented, or remove it.